lib/Common/Memory/PageAllocator.cpp

```diff
@@ -56,7 +57,20 @@ SegmentBase<T>::~SegmentBase()
         GetAllocator()->GetVirtualAllocator()->Free(originalAddress, GetPageCount() * AutoSystemInfo::PageSize, MEM_RELEASE, this->GetAllocator()->processHandle);
         GetAllocator()->ReportFree(this->segmentPageCount * AutoSystemInfo::PageSize); //Note: We reported the guard pages free when we decommitted them during segment initialization
 #if defined(_M_X64_OR_ARM64) && defined(RECYCLER_WRITE_BARRIER_BYTE)
-        RecyclerWriteBarrierManager::OnSegmentFree(this->address, this->segmentPageCount);
+#if ENABLE_DEBUG_CONFIG_OPTIONS
+        if (Js::Configuration::Global.flags.StrictWriteBarrierCheck)
+        {
+            if (this->isWriteBarrierEnabled)
+            {
+                RecyclerWriteBarrierManager::OnSegmentFree(this->address, this->segmentPageCount);
+                RecyclerWriteBarrierManager::ToggleBarrier(this->address, this->segmentPageCount * AutoSystemInfo::PageSize, false);
+            }
+        }
+        else
+#endif
+        {
+            RecyclerWriteBarrierManager::OnSegmentFree(this->address, this->segmentPageCount);
+        }
```

Comment:
Looks this fragment can be simplified a bit. Always call "OnSegmentFree". After that, do "#if...if... call ToggleBarrier". Avoids duplicate code "OnSegmentFree".